Add Feeditem class to replace the item array

logmanoriginal opened the pull request:
This PR adds a new class `FeedItem` to replace the current `item` array generated by the bridges. It is intended to perform sanity checks during bridge execution rather than attempting to format arbitrary contents (i.e. detecting invalid URIs and broken timestamps early).

* See `lib/FeedItem.php` for implementation details (documented in the source code)
* See `tests/FeedItemTest.php` for test cases (in case code breaks in the future)
* See any bridge for recommended changes (not required, because `index.php` still handles legacy arrays)

**Remarks**: 
- `index.php` currently supports `FeedItem` and legacy items. Updating bridges is not a requirement, but it is part of this PR right now.
- This PR updates all bridges to make use of the `$item->parameter` syntax (i.e. `$item->title` instead of `$item['title']`).
- `FeedItem` supports two methods for accessing parameters:
  - `$item->setParameter(...)`, for example `$item->setTitle($title)` and
  - `$item->parameter = ...`, for example `$item->title = $title`, at the cost of performance according to https://stackoverflow.com/a/3330859 (though I think it can be neglected because downloading and caching is the real bottleneck)
- This PR might break some bridges which have send invalid data in the first place (for example because the item URI (`$item['uri']`) is relative instead of absolute (which it should be).

Let me know if there is any problem with this approach.
Are there any arguments against updating all bridges?

logmanoriginal committed:
core: Add FeedItem to replace the legacy item array

logmanoriginal committed:
phpunit: Add unit tests for FeedItem

logmanoriginal committed:
[index] Transform legacy items to FeedItems if necessary

Add transformation from legacy items to FeedItems, before transforming
items to the desired format. This allows using legacy bridges alongside
bridges that return FeedItems.

logmanoriginal committed:
format: Refactor classes to use FeedItem

logmanoriginal committed:
Fix broken __set function

This function did not function properly due to missing breaks
inside the switch statement.

logmanoriginal committed:
[FeedItem] Add support for simple_html_dom objects to setContent()

logmanoriginal committed:
[index] Fix empty feed causes error message

em92 commented:
The only things I see are disadvantages:
- any invalid title, timestamp etc will produce broken feed
- cost of performance

>  at the cost of performance according to https://stackoverflow.com/a/3330859 (though I think it can be neglected because downloading and caching is the real bottleneck)

logmanoriginal commented:
Thanks for the feedback!
>* any invalid title, timestamp etc will produce broken feed

Those bridges currently fail as well, but silently over long periods of time until a user realizes something is missing (i.e. missing titles, no timestamps, etc...). It should be possible to detect such cases early on, in order to keep bridges as functional as possible.

I do agree that forcefully breaking bridges is not a good solution, especially because it breaks user experience (which I hadn't considered tbh). Debug messages are a better approach to this, as they are disabled in production mode. I'll work on that.

>* cost of performance

That only applies to the `$item->parameter` syntax (which is simpler to use). As mentioned before, this is probably no issue for RSS-Bridge, because downloading contents takes much more performance. I haven't seen any differences in performance on my machine.

Anyway, I'll revert the last commit (changing all bridges) for now, this can be managed more efficiently in `index.php` using the more efficient setter functions. Performance isn't an issue then and we can still upgrade bridges later.

***

>The only things I see are disadvantages

Here are a few advantages:

* **Predictability**
Parameters are specified explicitly and not implicitly like before. This makes things more predictable and (hopefully) helps new developers in implementing their own bridges with better feedback on what had gone wrong (no more empty output). Invalid parameters are properly reported and no longer ignored (i.e. relative URIs, empty titles, etc...). Currently those invalid parameters are only noticed by looking at the output, which may degrade over time.

* **Simplicity**
Formats can be simplified because they don't need to validate parameters (i.e. Enclosures and Categories are guaranteed to be arrays, Title will be text, etc...). Those problems are sometimes hard to detect and should not result in a broken formats, as they indicate degrading or broken bridges.

* **Standardization**
Bridges don't need to check if parameters are set. 
Code like `if(!isset($item['enclosures'])) { $item['enclosures'] = array(); }` is no longer necessary.

* **Extensibility**
FeedItem could handle common operations (to be fair, these are not implemented yet). Those common operations will benefit all bridges, without the need to touch any of them. Here are a few examples:
    - ~If timestamp is a string, try `strtotime`~
    - ~Do intelligent parsing if a parameter is an instance of `simple_html_dom`. For example, if URI is such an object, try `$uri->href`.~
    - ~Remove script tags and forms from `content`~
    - Strip tags from titles, but keep it readable (i.e. put spaces where necessary)
    - Take care of entity decoding (especially for `content`)

***

Forgot to mention: caching doesn't work yet.

logmanoriginal committed:
[FeedItem] Don't throw exceptions

As discussed in #940, instead of throwing exceptions on invalid
parameters, add messages to the debug log instead

logmanoriginal committed:
[FeedItem] Fix broken construct from item

PHP doesn't use the __set() function when using the $object->parameter
syntax. __set() must be called directly inside classes.

logmanoriginal committed:
Revert "phpunit: Add unit tests for FeedItem"

This reverts commit 1237b7bea486de84c3222018ebb23f9af57ca6fa.

The unit tests don't work if no exceptions are thrown by the class.

logmanoriginal committed:
[FeedItem] Fix timestamp detection

This fixes timestamp detection where numeric strings are
incorrectly flagged as non-integer when using is_int.

logmanoriginal committed:
[FeedItem] Support strings for setTimestamp using strtotime

Add support for strings to setTimestamp(). If the provided timestamp
is a string, automatically try to parse it using strtotime().

This allows bridges to simply use `$item['timestamp'] = $timestamp;`
instead of `$item['timestamp'] = strtotime($timestamp);`

logmanoriginal committed:
[FeedItem] Properly ignore broken values

logmanoriginal committed:
[FeedItem] Support simple_html_dom_node as input paramter for setURI

logmanoriginal committed:
[FeedItem] Support simple_html_dom_node as input parameter for setContent

logmanoriginal committed:
[index] Fix caching

logmanoriginal committed:
[index] Fix exception messages don't generate feed items

triatic commented:
This seems to be triggering deprecation notices on PHP 7.3, raised as #981 .
